### PR TITLE
DDF-1355: Default Geocoder using the local Lucene index

### PR DESCRIPTION
**Pull request by jrnorth**
[Review on Reviewable](https://reviewable.io/reviews/codice/ddf-catalog/134)

### CODE REVIEW

**Comment by jrnorth** [comment]
@millerw8 @jlcsmith @kcwire @wmcnalli @beyelerb @emanns95 @kguerriero @shane-voisard

File: ui/search-ui/search-geonames/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -19,4 +19,12 @@
 
     <service ref="geoCoder" interface="org.codice.ddf.ui.searchui.geocoder.GeoCoder"/>
 
+    <bean id="geoCoderLocal" class="org.codice.ddf.ui.searchui.geocoder.geonames.GeoNamesLocalIndex">
+        <property name="geoEntryQueryable" ref="geoEntryQueryable"/>
+    </bean>
+
+    <reference id="geoEntryQueryable" interface="org.codice.ddf.spatial.geocoding.GeoEntryQueryable"/>
+
+    <service ref="geoCoderLocal" interface="org.codice.ddf.ui.searchui.geocoder.GeoCoder"/>
```
Comment (jrnorth) [inline ui/search-ui/search-geonames/src/main/resources/OSGI-INF/blueprint/blueprint.xml:28]: Without specifying rankings, I'm not sure which of the two services (online or offline) will be used by default.

**Comment by jaymcnallie** [comment]
Review status: 0 of 25 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

<sup>**[ui/search-ui/search-geonames/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 28 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvqIQzdLaHDiYiYCCqx-r2-28)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/ui/search-ui/search-geonames/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L28)):</sup>
Could we leverage blueprint factory-method coupled with a metatype to make it configurable and then just fire up local or remote GeoCoder instance based on the config?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/134)

**Comment by jaymcnallie** [comment]
Reviewed 22 of 24 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/134)

**Comment by kcwire** [comment]
Reviewed 22 of 24 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

<sup>**[spatial/geocoding/spatial-geocoding-index/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesLuceneIndexer.java, line 60 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-Jvyin5bIAnAdq9IHDZT)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-index/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesLuceneIndexer.java#L60)):</sup>
Probably worth a comment as to why you are doing this.

---

<sup>**[spatial/geocoding/spatial-geocoding-index/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesLuceneIndexer.java, line 196 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvyilCPDW50Ij3hLOno)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-index/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesLuceneIndexer.java#L196)):</sup>
What's ADM?

---

<sup>**[spatial/geocoding/spatial-geocoding-index/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesLuceneIndexer.java, line 201 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-Jvyiw5YoprHb0YvFiP8)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-index/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesLuceneIndexer.java#L201)):</sup>
There's a lot of magic numbers in here - you probably want to add a comment so the next person understands

---

<sup>**[spatial/geocoding/spatial-geocoding-query/src/main/java/org/codice/ddf/spatial/geocoding/query/GeoNamesQueryLuceneIndex.java, line 101 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvyjWL9gMvEX3w9m_Bg)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-query/src/main/java/org/codice/ddf/spatial/geocoding/query/GeoNamesQueryLuceneIndex.java#L101)):</sup>
if significant make it a constant.

---

<sup>**[spatial/geocoding/spatial-geocoding-query/src/main/java/org/codice/ddf/spatial/geocoding/query/GeoNamesQueryLuceneIndex.java, line 103 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvyjbmFBEWM-jP04Dxk)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-query/src/main/java/org/codice/ddf/spatial/geocoding/query/GeoNamesQueryLuceneIndex.java#L103)):</sup>
It's probably worth adding all these field names as constants somewhere. maybe add a GeoCodingConstants.

---

<sup>**[spatial/geocoding/spatial-geocoding-query/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 22 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvyjvC0jjnCmuudpEz9)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-query/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L22)):</sup>
Can you confirm that you can create the directory outside the DDF_HOME dir? There could be cases where you want this index to be maintained outside the distro

---

<sup>**[ui/search-ui/search-geonames/src/main/java/org/codice/ddf/ui/searchui/geocoder/geonames/GeoNamesLocalIndex.java, line 1 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-Jvykx8eszXeDj-P4uva)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/ui/search-ui/search-geonames/src/main/java/org/codice/ddf/ui/searchui/geocoder/geonames/GeoNamesLocalIndex.java#L1)):</sup>
I would not add this implementation to the same package/project as another implementation.  This is an alternative and should be independent of the GeoNames Webservice implementation.

---

<sup>**[ui/search-ui/search-geonames/src/main/java/org/codice/ddf/ui/searchui/geocoder/geonames/GeoResultCreator.java, line 28 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvylHoIMWicql1V3zM2)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/ui/search-ui/search-geonames/src/main/java/org/codice/ddf/ui/searchui/geocoder/geonames/GeoResultCreator.java#L28)):</sup>
you do have constants!

---

<sup>**[ui/search-ui/search-geonames/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 28 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvqIQzdLaHDiYiYCCqx-r2-28)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/ui/search-ui/search-geonames/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L28)):</sup>
I would think this belongs in a separate project and a separate bundle.  It's a bit odd to have 2 implementations in the same project.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/134)

**Comment by jrnorth** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

<sup>**[spatial/geocoding/spatial-geocoding-query/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 22 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-JvyjvC0jjnCmuudpEz9)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/spatial/geocoding/spatial-geocoding-query/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L22)):</sup>
You can if you specify an absolute path.

---

<sup>**[ui/search-ui/search-geonames/src/main/java/org/codice/ddf/ui/searchui/geocoder/geonames/GeoNamesLocalIndex.java, line 1 [r2]](https://reviewable.io:443/reviews/codice/ddf-catalog/134#-Jvykx8eszXeDj-P4uva)** ([raw file](https://github.com/codice/ddf-catalog/blob/00b348d47c6d5ec5c751b0e97c5139d2f3b51366/ui/search-ui/search-geonames/src/main/java/org/codice/ddf/ui/searchui/geocoder/geonames/GeoNamesLocalIndex.java#L1)):</sup>
Should the new project also reside in ui or somewhere else?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/134)

**Pull request closed without merging**